### PR TITLE
use an upsert to update device_lists_outbound_last_success

### PR DESCRIPTION
**Pull request by richvdh**
I rather hope that this will be more efficient than the current round-the-houses thing it does.

### CODE REVIEW

File: synapse/storage/data_stores/main/devices.py [hunk 1]
```diff
@@ -725,6 +720,21 @@ def __init__(self, database: Database, db_conn, hs):
...
+        # create a unique index on device_lists_outbound_last_success
+        self.db.updates.register_background_index_update(
+            "device_lists_outbound_last_success_unique_idx",
+            index_name="device_lists_outbound_last_success_unique_idx",
+            table="device_lists_outbound_last_success",
+            columns=["destination", "user_id"],
+            unique=True,
+        )
+
+        # once that completes, we can remove the old non-unique indexes.
```
Comment (babolivier) [inline synapse/storage/data_stores/main/devices.py:732]:
```suggestion
        # once that completes, we can remove the old non-unique index.
```

Looks like we're only dropping one index.

File: synapse/storage/data_stores/main/devices.py [hunk 2]
```diff
@@ -799,6 +809,20 @@ def _txn(txn):
 
         return rows
 
+    async def _drop_device_lists_outbound_last_success_non_unique_idx(
+        self, progress, batch_size
+    ):
+        def f(conn):
+            txn = conn.cursor()
+            txn.execute("DROP INDEX IF EXISTS device_lists_outbound_last_success_idx")
+            txn.close()
```
Comment (babolivier) [inline synapse/storage/data_stores/main/devices.py:818]: Isn't this function missing a `conn.commit()`? Also, why not run it with `runInteraction` so we just get a cursor and don't need to care about the connection?

File: synapse/storage/database.py [hunk 3]
```diff
@@ -889,20 +891,24 @@ def simple_upsert_txn_native_upsert(
...
+        txn: LoggingTransaction,
+        table: str,
+        key_names: Collection[str],
+        key_values: Collection[Iterable[Any]],
+        value_names: Collection[str],
+        value_values: Iterable[Iterable[str]],
+    ) -> None:
         """
-        Upsert, many times.
+        Upsert, many times, but without native UPSERT support or batching.
```
Comment (babolivier) [inline synapse/storage/database.py:903]: > but without native UPSERT support or batching

I didn't dig through it too much, but the fact that this function does is call `simple_upsert_many_txn_native_upsert` (when the engine supports it), which mentions "using batching where possible", feels like this doc is lying.

Comment (richvdh) [reply synapse/storage/database.py:903]: oops, this was a bad c&P, thanks

File: synapse/storage/data_stores/main/devices.py (hunk 2 above)
Comment (richvdh) [reply synapse/storage/data_stores/main/devices.py:818]: I dunno. I just cargo-culted this from similar code elsewhere.

I *think* it's something to do with schema-changing commands having weird interactions with cursors/transactions. but I really don't know.

Comment (richvdh) [reply synapse/storage/data_stores/main/devices.py]: right, I did some digging. It seems to have originated with https://github.com/matrix-org/synapse/commit/5a32e9273ec9759caf09d5b8204dd29e7a007b97 - or rather a predecessor to that change. As far as I can see there is no reason for it to still be done this way, so I've changed it.

**Pull request merged**